Web Inspector: New CSS property unexpectedly dropped from empty CSS rule when tabbing through or editing selector

rcaliman-apple opened the pull request:
#### 84150ea85574551811c8802fe27fa252bc571d31
<pre>
Web Inspector: New CSS property unexpectedly dropped from empty CSS rule when tabbing through or editing selector
<a href="https://bugs.webkit.org/show_bug.cgi?id=245768">https://bugs.webkit.org/show_bug.cgi?id=245768</a>

Reviewed by Patrick Angle.

There are two issues here:
1) changing the selector of an empty CSS rule causes the first added CSS property to be dropped.
2) tabbing through the editable text field for a CSS selector marks it as changed even when it&apos;s not.

The core problem is the order of operations.

Tabbing out of a CSS selector text field first calls
`SpreadsheetCSSStyleDeclarationSection.spreadsheetRuleHeaderFieldWillNavigate()`
from the `SpreadsheetRuleHeaderField._handleKeyDown` handler.

For an empty CSS rule, this creates a new blank CSS property with:
`SpreadsheetCSSStyleDeclarationEditor.startEditingFirstProperty()` -&gt; `CSSStyleDeclaration.newBlankProperty()`.

At this point, the frontend model for the CSS rule contains a `CSSStyleDeclaration` with one `CSSProperty`.

In quick succession, the second operation happens when the selector text field loses focus:
`SpreadsheetCSSStyleDeclarationSection.spreadsheetRuleHeaderFieldDidCommit()`
from the `SpreadsheetRuleHeaderField._handleBlur` handler.

If the selector text has changed, a request is sent to the backend via `DOMNodeStyles.changeRuleSelector()`.
A request is then made for the latest matching styles via `DOMNodeStyles.refresh()`.

For an empty CSS rule, there are no matching styles. As a result, the `CSSStyleDeclaration` of the corresponding
CSS rule is updated with an empty list of CSS properties in `DOMNodeStyles._parseStyleDeclarationPayload()`:

```
if (style) {
    style.update(text, properties, styleSheetTextRange);
    return style;
}
```

This has the effect of orphaning the new blank `CSSProperty` introduced earlier.

`CSSProperty._updateOwnerStyleText()` is called to persist the `CSSProperty` to the backend.
This calls `CSSStyleDeclaration.generateFormattedText()` to serialize the whole CSS declaration block.
But our `CSSStyleDeclaration` was just emptied of properties by the update in
`DOMNodeStyles._parseStyleDeclarationPayload()` so it falls back to returning an empty string:

```
let styleText = &quot;&quot;;
...
let properties = this._styleSheetTextRange ? this.visibleProperties : this._properties;
...
return styleText;
```

When the `CSSProperty` commits with an empty name and value, the matching styles obtained
in response to `DOMNodeStyles.refresh()` are empty, and the orphaned property is dropped.

Trying again with the newly created blank CSS property will work because this time
there&apos;s no race with `DOMNodeStyles.refresh()` from the selector change.

To fix the first issue, we must ensure that the operation to update the selector
and the style refresh it causes occurs before adding a new CSS property to avoid
the style refresh invalidating the models on the frontend.

To fix the second issue, we don&apos;t call `SpreadsheetRuleHeaderField.stopEditing()` in the
keydown handler for Tab or Enter  because that resets `_valueBeforeEditing` which causes
the check in  `SpreadsheetRuleHeaderField._handleBlur` to always consider the selector
as having been changed, thus triggering the operations described above.

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldWillNavigate):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
(WI.SpreadsheetRuleHeaderField):
(WI.SpreadsheetRuleHeaderField.prototype.stopEditing):
(WI.SpreadsheetRuleHeaderField.prototype._handleBlur):
(WI.SpreadsheetRuleHeaderField.prototype._handleKeyDown):

Canonical link: <a href="https://commits.webkit.org/261861@main">https://commits.webkit.org/261861@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e3bd753a16a4129070a6acbc77ab4c216153434

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112919 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22077 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1594 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4695 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121416 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23424 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13238 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5883 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118706 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17395 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100650 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106008 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99355 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1188 "5 failures") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46417 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14372 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1228 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10563 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20388 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53220 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8282 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16928 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -748,15 +749,15 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
     {
         if (direction === "forward")
-            this._propertiesEditor.startEditingFirstProperty();
+            this._setSelectorTextPromise.then(() => this._propertiesEditor.startEditingFirstProperty());
```
Having UI interactions like tabbing wait on the backend seems like it will create a poor experience when trying to quickly tab through items. Doesn't your other PR fix the race that adding a new rule gets removed when refreshing from the backend. And also doesn't the lock mechanism on CSSProperty prevent the newly created property from being deleted while its being edited?

rcaliman-apple replied:
> Having UI interactions like tabbing wait on the backend seems like it will create a poor experience when trying to quickly tab through items.

`_setSelectorTextPromise` is instantiated with a resolved promise in the constructor:

```
this._setSelectorTextPromise = Promise.resolve();
```

This means that tabbing through selector fields without changing them will be fast. No trip to the backend. The bug where tabbing through fields always marked them as changed was fixed in this patch in `WI.SpreadsheetRuleHeaderField._handleKeyDown()` by removing the call to `stopEditing()` and letting the blur event handler call that.

If a selector is intentionally edited while tabbing, the promise of sending it to the backend is assigned to `_setSelectorTextPromise`. This will be awaited before moving the focus to the first property. That seems reasonable. Once that resolves, it remains as a resolved promise so it won't slow down additional tabbing.

> Doesn't your other PR fix the race that adding a new rule gets removed when refreshing from the backend.

That's a different behaviour: adding a new CSS rule. Here we're dealing with adding a new **CSS property** to an empty rule that's either already present in the stylesheet or already written to the inspector stylesheet.

The race condition fixed in #11547 there doesn't solve this issue.

> And also doesn't the lock mechanism on CSSProperty prevent the newly created property from being deleted while its being edited?

I was intrigued by this comment so I investigated further.

The lock is set on `CSSStyleDeclaration` when there's focus within an element (text field or inline swatch) in 
[`SpreadsheetCSSStyleDeclarationEditor.initialLayout()`](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationEditor.js#L63-L70):

```js
this.element.addEventListener("focus", () => { this.focused = true; }, true);
this.element.addEventListener("blur", (event) => {
    let focusedElement = event.relatedTarget;
    if (focusedElement && this.element.contains(focusedElement))
        return;

    this.focused = false;
}, true);
```

The selector text field is not a child of a `SpreadsheetCSSStyleDeclarationEditor` so it doesn't trigger those "focus" and "blur" event handlers, but that's not important here, just odd.

The lock _could_ work, but the current check in [`CSSStyleDeclaration.update()`](https://github.com/WebKit/WebKit/blob/3fabfea068b81f41e1c30cde8743a744aaa12dc9/Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js#L183-L184) doesn't catch the conditions for this bug because `this._text` and the incoming `text` are both empty strings as soon as the property is created. The selector text field blurs too quickly for the user to have a chance to type anything in the property name field.

```js
if (this._locked && !options.forceUpdate && text !== this._text)
    return;
```

It would need to be changed:

```diff
- if (this._locked && !options.forceUpdate && text !== this._text)
+ let isEmpty = (this._text && this._text.trim() === "") && text.trim() === ""; // `this._text` can be `undefined`
+ if (this._locked && !options.forceUpdate && (text !== this._text || isEmpty))
```

This _does work_, but it feels like a hack, honestly.

The big issue I see in this bug is that the order of operations is wrong: the blank property gets added before committing the changed selector text so the refresh of matching styles ends up clobbering the list of properties. The check with the `CSSStyleDeclaration._lock` is trying guard against clobbering of various edge-cases.

I'm more satisfied with the approach in the existing patch where the order of operations is corrected and the style refresh from the selector change is explicitly awaited before adding a new blank css property. The code is easier to reason about and truthful about its intent.

This bug was subtle and took a long time to debug. The clobbering happens silently. I'm hesitant to add further to the condition for the early return in `CSSStyleDeclaration.update()` out of concern for adding more subtle bugs.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
```diff
@@ -41,6 +41,7 @@ WI.SpreadsheetRuleHeaderField = class SpreadsheetRuleHeaderField extends WI.Obje
         this._editing = false;
         this._valueBeforeEditing = "";
         this._handledMouseDown = false;
+        this._navigateDirection = null;
```
Maybe `tabNavigationDirection` better communicates what this variable does?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
```diff
@@ -116,6 +118,10 @@ WI.SpreadsheetRuleHeaderField = class SpreadsheetRuleHeaderField extends WI.Obje
 
+        if (this._delegate?.spreadsheetRuleHeaderFieldWillNavigate && this._navigateDirection) {
+            this._delegate.spreadsheetRuleHeaderFieldWillNavigate(this, this._navigateDirection);
+        }
```
we can drop the `this._delegate` checks and just have it only check for `this._navigateDirection` since ultimately that's all that matters.  IMO we only really need to check the `this._delegate` if we're observing the `return` or using it as a "dont do the regular path"
```
if (this._navigateDirection)
    this._delegate?.spreadsheetRuleHeaderFieldWillNavigate?.(this, this._navigateDirection);
```

also, i wonder if we'd be able to look at anything on the `event` to see the cause of the `"blur"` instead of having to have another member variable 🤔

Style: no `{` `}` for single line control flow

rcaliman-apple replied:
>also,  i wonder if we'd be able to look at anything on the event to see the cause of the "blur" instead of having to have another member variable

I tried manually dispatching a "blur" event but can't modify the `detail` option to hold an arbitrary value.
[`UIEvent.detail`](https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/detail) is supposed to hold a number and has special meaning. I'd rather not mess with that. Dispatching a mock "blur" event as a `CustomEvent` is just asking for trouble. The extra member will have to do.

dcrousso reviewed (commented):
> If the selector text has changed, a request is sent to the backend via `DOMNodeStyles.changeRuleSelector()`.
> A request is then made for the latest matching styles via `DOMNodeStyles.refresh()`.

what about if the selector text hasnt changed?  id hope that we would bail pretty quickly out of any work

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
```diff
@@ -111,10 +113,11 @@ WI.SpreadsheetRuleHeaderField = class SpreadsheetRuleHeaderField extends WI.Obje
-            this._delegate.spreadsheetRuleHeaderFieldDidCommit(this, changed);
-        }
+        let changed = this._element.textContent !== this._valueBeforeEditing;
+        this._delegate.spreadsheetRuleHeaderFieldDidCommit(this, changed);
```
Based on the previous condition for calling this, shouldn't we use optional chaining here since both `_delegate` as well as `spreadsheetRuleHeaderFieldDidCommit` are optional?
e.g. `this._delegate?.spreadsheetRuleHeaderFieldDidCommit?.(this, changed);`

rcaliman-apple replied:
🤦 Indeed. I took Devin's suggestion but forgot to move the checks. Will update patch.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
```diff
@@ -111,10 +113,11 @@ WI.SpreadsheetRuleHeaderField = class SpreadsheetRuleHeaderField extends WI.Obje
+        this._delegate.spreadsheetRuleHeaderFieldDidCommit(this, changed);
+
+        if (this._tabNavigationDirection)
+            this._delegate.spreadsheetRuleHeaderFieldWillNavigate(this, this._tabNavigationDirection);
```
Ditto :117

rcaliman-apple committed:
Web Inspector: New CSS property unexpectedly dropped from empty CSS rule when tabbing through or editing selector
https://bugs.webkit.org/show_bug.cgi?id=245768

Reviewed by Patrick Angle.

There are two issues here:
1) changing the selector of an empty CSS rule causes the first added CSS property to be dropped.
2) tabbing through the editable text field for a CSS selector marks it as changed even when it's not.

The core problem is the order of operations.

Tabbing out of a CSS selector text field first calls
`SpreadsheetCSSStyleDeclarationSection.spreadsheetRuleHeaderFieldWillNavigate()`
from the `SpreadsheetRuleHeaderField._handleKeyDown` handler.

For an empty CSS rule, this creates a new blank CSS property with:
`SpreadsheetCSSStyleDeclarationEditor.startEditingFirstProperty()` -> `CSSStyleDeclaration.newBlankProperty()`.

At this point, the frontend model for the CSS rule contains a `CSSStyleDeclaration` with one `CSSProperty`.

In quick succession, the second operation happens when the selector text field loses focus:
`SpreadsheetCSSStyleDeclarationSection.spreadsheetRuleHeaderFieldDidCommit()`
from the `SpreadsheetRuleHeaderField._handleBlur` handler.

If the selector text has changed, a request is sent to the backend via `DOMNodeStyles.changeRuleSelector()`.
A request is then made for the latest matching styles via `DOMNodeStyles.refresh()`.

For an empty CSS rule, there are no matching styles. As a result, the `CSSStyleDeclaration` of the corresponding
CSS rule is updated with an empty list of CSS properties in `DOMNodeStyles._parseStyleDeclarationPayload()`:

```
if (style) {
    style.update(text, properties, styleSheetTextRange);
    return style;
}
```

This has the effect of orphaning the new blank `CSSProperty` introduced earlier.

`CSSProperty._updateOwnerStyleText()` is called to persist the `CSSProperty` to the backend.
This calls `CSSStyleDeclaration.generateFormattedText()` to serialize the whole CSS declaration block.
But our `CSSStyleDeclaration` was just emptied of properties by the update in
`DOMNodeStyles._parseStyleDeclarationPayload()` so it falls back to returning an empty string:

```
let styleText = "";
...
let properties = this._styleSheetTextRange ? this.visibleProperties : this._properties;
...
return styleText;
```

When the `CSSProperty` commits with an empty name and value, the matching styles obtained
in response to `DOMNodeStyles.refresh()` are empty, and the orphaned property is dropped.

Trying again with the newly created blank CSS property will work because this time
there's no race with `DOMNodeStyles.refresh()` from the selector change.

To fix the first issue, we must ensure that the operation to update the selector
and the style refresh it causes occurs before adding a new CSS property to avoid
the style refresh invalidating the models on the frontend.

To fix the second issue, we don't call `SpreadsheetRuleHeaderField.stopEditing()` in the
keydown handler for Tab or Enter  because that resets `_valueBeforeEditing` which causes
the check in  `SpreadsheetRuleHeaderField._handleBlur` to always consider the selector
as having been changed, thus triggering the operations described above.

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldWillNavigate):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
(WI.SpreadsheetRuleHeaderField):
(WI.SpreadsheetRuleHeaderField.prototype.stopEditing):
(WI.SpreadsheetRuleHeaderField.prototype._handleBlur):
(WI.SpreadsheetRuleHeaderField.prototype._handleKeyDown):

Canonical link: https://commits.webkit.org/261861@main
